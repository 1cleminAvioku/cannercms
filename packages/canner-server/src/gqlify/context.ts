import { Context } from 'koa';
import Boom from 'boom';
import { ErrorCodes } from './error';
import { GqlifyConfig } from './config';
import { isNil, get } from 'lodash';
import { createAuthHandler } from '../common/graphql';

export const createContext = (config: GqlifyConfig) => {
  const publicAccess = isNil(config.oidc);

  if (publicAccess) {
    return null;
  }

  const authHandler = createAuthHandler({
    discoveryUrl: get(config, 'oidc.discoveryUrl'),
    issuerConfig: get(config, 'oidc.issuerConfig'),
    readOnlyAccessToken: config.readOnlyAccessToken,
  });

  return async ({ ctx }: { ctx: Context }) => {
    const authorization = ctx.get('authorization');
    if (!authorization) {
      throw Boom.forbidden('authorization header required', {code: ErrorCodes.AUTH_HEADER_MISSING});
    }

    const accessToken = authorization.replace('Bearer ', '');
    if (!accessToken) {
      throw Boom.forbidden('access token required', {code: ErrorCodes.ACCESSTOKEN_MISSING});
    }

    // check with authHandler
    const authContext = await authHandler(accessToken);

    return {
      ...authContext
    };
  };
};
