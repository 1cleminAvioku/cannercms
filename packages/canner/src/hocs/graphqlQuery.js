import * as React from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';

export default function withQuery(Com) {
  return class ComWithQuery extends React.Component {
    render() {
      const {graphql, ...restProps} = this.props;
      return (
        <Query query={gql`${graphql}`} >
          {({loading, error, data}) => {
            if (loading) return null;
            if (error) return `Error!: ${error}`;
            const key = Object.keys(data)[0];
            return (
              <Com value={data[key].map(d => ({...d}))} {...restProps} />
            );
          }}
        </Query>
      );
    }
  }
}