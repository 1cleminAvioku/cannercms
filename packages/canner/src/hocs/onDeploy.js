// @flow

import * as React from 'react';
import {get, isArray, isPlainObject, set} from 'lodash';
import RefId from 'canner-ref-id';
import type {HOCProps} from './types';

export default function withOndeploy(Com: React.ComponentType<*>) {
  return class ComponentWithOnDeploy extends React.Component<HOCProps> {
    key: string;
    id: ?string;

    constructor(props: HOCProps) {
      super(props);
      const {pattern, refId, rootValue} = props;
      const {key, id} = splitRefId({
        refId,
        rootValue,
        pattern
      });
      this.key = key;
      this.id = id;
    }

    removeOnDeploy = (arg1: string, callbackId: string) => {
      const {removeOnDeploy} = this.props;
      if (callbackId) {
        return removeOnDeploy(arg1, callbackId);
      } else {
        return removeOnDeploy(this.key, arg1);
      }
    }

    onDeploy = (arg1: string | Function, callback: Function): string => {
      const {onDeploy, refId} = this.props;
      if (typeof arg1 === 'string') {
        return onDeploy(arg1, callback);
      } else {
        // first arguments is a function
        return onDeploy(this.key, result => {
          let restPathArr = refId.getPathArr();
          // if (this.id) {
          //   restPathArr = restPathArr.slice(2);
          // } else {
            restPathArr = restPathArr.slice(1);
          // }
          const {paths, value} = getValueAndPaths(result.data, restPathArr);
          return {
            ...result,
            // $FlowFixMe
            data: set(result.data, paths, arg1(value))
          }
        });
      }
    }

    render() {
      return <Com {...this.props}
        onDeploy={this.onDeploy}
        removeOnDeploy={this.removeOnDeploy}
      />
  }
  };
}


export function splitRefId({
  refId,
  rootValue,
  pattern
}: {
  refId: RefId,
  rootValue: any,
  pattern: string
}) {
  const [key, index] = refId.getPathArr();
  let id;
  if (pattern.startsWith('array')) {
    id = get(rootValue, [key, index, 'id']);
  }
  return {
    key,
    id
  }
}

export function getValueAndPaths(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    let v = result.value;
    let paths = result.paths;
    if (isPlainObject(v)) {
      if ('edges' in v && 'pageInfo' in v) {
        v = get(v, ['edges', key, 'node']);
        paths = paths.concat(['edges', key, 'node']);
      } else {
        v = get(v, key);
        paths = paths.concat(key);
      }
    } else if (isArray(v)) {
      v = get(v, key);
      paths = paths.concat(key);
    }
    return {
      value: v,
      paths
    }
  }, {
    value,
    paths: []
  });
}