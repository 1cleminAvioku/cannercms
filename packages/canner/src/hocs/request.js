// @flow

import * as React from 'react';
import {generateAction} from '../action';
import {createEmptyData} from 'canner-helpers';
import RefId from 'canner-ref-id';
import type {HOCProps} from './types';
import {update, merge, isArray, isPlainObject, mapValues} from 'lodash';

type changeQueue = Array<{RefId: RefId | {firstRefId: RefId, secondRefId: RefId}, type: any, value: any}>;

export default function withRequest(Com: React.ComponentType<*>) {
  // this hoc will update data;
  return class ComponentWithRequest extends React.Component<HOCProps> {
    onChange = (refId: RefId | {firstRefId: RefId, secondRefId: RefId} | changeQueue, type: any, delta: any, config: any): Promise<*> => {
      let id;
      if (isArray(refId)) { // changeQueue
        const changeQueue = refId;
        // $FlowFixMe
        return Promise.all(changeQueue.map(args => {
          const {refId, type, value} = args;
          return this.onChange(refId, type, value);
        }));
      } else if (refId instanceof RefId) {
        id = refId.toString();
      } else {
        id = {
          // $FlowFixMe: refId should be object here
          firstId: refId.firstRefId.toString(),
          // $FlowFixMe: refId should be object here
          secondId: refId.secondRefId.toString()
        };
      } 
      const {schema, rootValue, request} = this.props;
      const itemSchema = findSchemaByRefId(schema, refId);
      const {relation, items, pattern} = itemSchema;
      const action = createAction({
        relation,
        id,
        type,
        value: delta,
        config,
        rootValue,
        items,
        pattern
      });
      if (!action) {
        throw new Error('invalid change');
      }
      if (action.type === 'NOOP') {
        return Promise.resolve();
      }
      // $FlowFixMe
      return request(action);
    }

    render() {
      return <Com {...this.props} onChange={this.onChange} />;
    }
  };
}

export function createAction({
  relation,
  id,
  type,
  value,
  config,
  rootValue,
  items,
}: { 
  relation: any,
  id: {firstId: string, secondId: string} | string,
  type: 'create' | 'update' | 'delete' | 'swap',
  value: any,
  config: any,
  rootValue: any,
  value: any,
  items: any
}) {
  if (type === 'create') {
    if (!config) {
      const emptyData = createEmptyData(items);
      if (isPlainObject(emptyData)) {
        value = merge(emptyData, value);
      } else {
        value = emptyData;
      }
    }
    if (typeof id === 'string' && id.split('/').length === 1) {
      update(value, 'id', id => id || randomId());
    }
  }
  value = addTypename(value)
  return generateAction({
    id,
    updateType: type,
    value,
    rootValue,
    relation
  });
}

function randomId() {
  return Math.random().toString(36).substr(2, 12);
}

function addTypename(value) {
  if (isPlainObject(value)) {
    update(value, '__typename', typename => typename || null);
    value = mapValues(value, child => addTypename(child));
  } else if (isArray(value)) {
    value = value.map(child => addTypename(child));
  }
  return value;
}

export function findSchemaByRefId(schema: Object, refId: any) {
  let paths = [];
  if (isPlainObject(refId)) {
    paths = refId.firstRefId.getPathArr();
  } else {
    paths = refId.getPathArr();
  }
  let pattern = '';
  return loop(schema, paths, pattern);
}

export function loop(schema: Object, paths: Array<string>, pattern: string): any {
  if (paths.length === 0) {
    return {
      ...schema,
      pattern: removeFirstSlash(pattern)
    };
  }

  if (!schema.type) { // in items of object
    return loop(schema[paths[0]], paths.slice(1), `${pattern}/${schema[paths[0]].type}`);
  }

  if (schema.type === 'json') {
    return {};
  }

  if (schema.type === 'array') {
    if (paths.length === 1) {
      // paths = [index], so just return
      return loop(schema, [], pattern);
    } else if (schema.items.type === 'object') {
      // path[0] is index, so we take the paths[1]
      return loop(schema.items.items[paths[1]], paths.slice(2), `${pattern}/${schema.items.items[paths[1]].type}`);
    }
  }
  return loop(schema.items[paths[0]], paths.slice(1), `${pattern}/${schema.items[paths[0]].type}`);
}

function removeFirstSlash(pattern: string) {
  return pattern.split('/').slice(1).join('/');
}