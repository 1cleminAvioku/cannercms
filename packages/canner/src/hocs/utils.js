// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}