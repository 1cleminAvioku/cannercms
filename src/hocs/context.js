// @flow
import * as React from 'react';

export const defaultContext = {
  componentId: '',
  query: {},
  fetch: () => {throw new Error('there is no fetch method');},
  subscribe: () => {throw new Error('there is no subscribe method');},
  request: () => {throw new Error('there is no request method');},
  deploy: () => {throw new Error('there is no deploy method');},
  reset: () => {throw new Error('there is not reset method');}
}

export const HOCContext = React.createContext(defaultContext);
