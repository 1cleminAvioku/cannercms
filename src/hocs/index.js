import query from './query';
import route from './route';
import cache from './cache';
import request from './request';
import deploy from './deploy';
import id from './id';
import title from './title';
import containerRouter from './containerRouter';
import onDeploy from './onDeploy';
import context from './connectContext';

export default {
  query,
  route,
  cache,
  request,
  deploy,
  // validator,
  id,
  title,
  containerRouter,
  onDeploy,
  context
};