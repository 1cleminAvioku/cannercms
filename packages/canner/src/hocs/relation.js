// @flow

import * as React from 'react';
import {Spin, Icon} from 'antd';
import RefId from 'canner-ref-id';
import Toolbar from './components/toolbar';
import {mapValues} from 'lodash';
import type {HOCProps} from './types';
import {parseConnectionToNormal, getValue, defaultValue} from './utils';
import {withApollo} from 'react-apollo';
import gql from 'graphql-tag';
import {Query} from '../query';
import {List} from 'react-content-loader';

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type State = {
  originRootValue: any,
  isFetching: boolean,
}

type Props = HOCProps & {
  client: any
}

@withApollo
export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    query: Query;

    constructor(props: Props) {
      super(props);
      this.state = {
        originRootValue: null,
        isFetching: true
      };
      if (props.relation) {
        this.query = new Query({schema: props.schema});
      }
    }

    componentDidMount() {
      // defaultSort
      const {relation, toolbar} = this.props;
      if (!relation) {
        return;
      }
      if (toolbar && toolbar.async) {
        const args = this.getArgs();
        this.updateQuery([relation.to], {
          ...args,
          first: 10
        });
      } else {
        this.queryData();
      }
      
    }

    UNSAFE_componentWillReceiveProps(props: Props) {
      const {refId, relation} = this.props;
      if (!relation) {
        return;
      }
      if (refId.toString() !== props.refId.toString()) {
        // refetch when route change
        this.queryData(props);
      }
    }

    queryData = (props?: Props): Promise<*> => {
      const {relation, client} = props || this.props;
      if (!relation) {
        return Promise.resolve();
      }
      this.setState({
        isFetching: true,
      });
      const gqlStr = this.query.toGQL(relation.to);
      const variables = this.query.getVairables();
      return client.query({
        query: gql`${gqlStr}`,
        variables
      }).then(({data}) => {
          this.setState({
            originRootValue: data,
            isFetching: false,
          });
        })
        .catch(() => {
          this.setState({
            isFetching: false
          })
        });
    }

    getArgs = () => {
      const {relation} = this.props;
      const queries = this.query.getQueries([relation.to]).args || {pagination: {first: 10}};
      const variables = this.query.getVairables();
      const args = mapValues(queries, v => variables[v.substr(1)]);
      return args;
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      this.query.updateQueries(paths, 'args', args);
      this.queryData();
    }

    render() {
      const {originRootValue, isFetching} = this.state;
      const {toolbar, relation, schema, refId} = this.props;
      if (!relation) {
        return <Com {...this.props}/>;
      }
      console.log(originRootValue);
      if (!originRootValue) {
        return <List style={{maxWidth: 500}} />;
      }
      const value = originRootValue[relation.to];
      const args = this.getArgs();
      const relationValue = value ? removeSelf(value, refId, relation.to) : defaultValue('connection');
      const tb = ({children, ...restProps}) => <Toolbar {...restProps}
        items={schema[relation.to].items.items}
        toolbar={toolbar || {pagination: {type: 'pagination'}}}
        args={args}
        query={this.query}
        keyName={relation.to}
        refId={new RefId(relation.to)}
        originRootValue={originRootValue}
        updateQuery={this.updateQuery}
        parseConnectionToNormal={parseConnectionToNormal}
        getValue={getValue}
        defaultValue={defaultValue}
      >
        <Spin indicator={antIcon} spinning={isFetching}>
          {children}
        </Spin>
      </Toolbar>;
      return <Com {...this.props} Toolbar={tb} relationValue={relationValue}/>;
    }
  };
}

export function removeSelf(value: any, refId: RefId, relationTo: string) {
  const [key, index] = refId.getPathArr().slice(0, 2);
  if (key !== relationTo) {
    return value;
  }
  return {...value, edges: value.edges.filter((v, i) => i !== Number(index))};
}
