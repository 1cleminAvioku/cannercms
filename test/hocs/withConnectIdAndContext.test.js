/**
|--------------------------------------------------
| this is just for copy and past
|--------------------------------------------------
*/


import * as React from 'react';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '../react163Adapter';
import withConnectId from '../../src/hocs/connectId';
import {HOCContext as Context} from '../../src/hocs/context';
import RefId from 'canner-ref-id';

Enzyme.configure({ adapter: new Adapter() });

describe('hocTemplate', () => {
  let WrapperComponent, props, MockComponent, context,
  mockFetch, mockSubscribe, mockRequest, mockDeploy, mockReset;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }
    props = {
      id: undefined,
      name: 'posts'
    }

    mockFetch = jest.fn();
    mockSubscribe = jest.fn();
    mockRequest = jest.fn();
    mockReset = jest.fn();
    mockDeploy = jest.fn();

    context = {
      componentId: undefined,
      query: undefined,
      fetch: mockFetch,
      subscribe: mockSubscribe,
      request: mockRequest,
      deploy: mockDeploy,
      reset: mockReset
    }

    WrapperComponent = withConnectId(MockComponent);
  });

  it('should render', () => {
    const wrapper = mount(
      <Context.Provider value={context}>
        <WrapperComponent
          {...props}
        />
      </Context.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should pass props', () => {
    const wrapper = mount(
      <Context.Provider value={context}>
        <WrapperComponent
          {...props}
        />
      </Context.Provider>
    );
    expect(wrapper.find(MockComponent).props()).toMatchObject({
      ...props,
      ...context,
      refId: expect.any(RefId),
      componentId: props.name,
      query: {}
    });
  });
});