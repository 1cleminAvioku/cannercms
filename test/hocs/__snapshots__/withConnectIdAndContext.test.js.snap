// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hocTemplate should render 1`] = `
<ComponentConnectIdAndContext
  name="posts"
>
  <MockComponent
    componentId="posts"
    deploy={[MockFunction]}
    fetch={[MockFunction]}
    name="posts"
    refId={
      CannerRefId {
        "_internalArr": Array [
          "posts",
        ],
        "child": [Function],
        "getPathArr": [Function],
        "remove": [Function],
        "toString": [Function],
      }
    }
    request={[MockFunction]}
    reset={[MockFunction]}
    subscribe={[MockFunction]}
  >
    <div>
      Component
    </div>
  </MockComponent>
</ComponentConnectIdAndContext>
`;
